src/argilla/cli/server/search_engine/reindex.py

```diff
@@ -0,0 +1,113 @@
+#  Copyright 2021-present, the Recognai S.L. team.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from typing import AsyncGenerator
+
+from rich.progress import Progress
+from sqlalchemy import func, select
+from sqlalchemy.ext.asyncio import AsyncSession
+from sqlalchemy.orm import selectinload
+
+from argilla.cli import typer_ext
+from argilla.server.database import AsyncSessionLocal
+from argilla.server.models import Dataset, Record, Response, Suggestion
+from argilla.server.search_engine import SearchEngine, get_search_engine
+
+
+class Reindexer:
+    YIELD_PER = 100
+
+    @classmethod
+    async def reindex_datasets(cls, db: AsyncSession, search_engine: SearchEngine) -> AsyncGenerator[Dataset, None]:
+        stream = await db.stream(
+            select(Dataset)
+            .order_by(Dataset.inserted_at.asc())
+            .options(
+                selectinload(Dataset.fields),
+                selectinload(Dataset.questions),
+                selectinload(Dataset.metadata_properties),
+                selectinload(Dataset.vectors_settings),
+            )
+            .execution_options(yield_per=cls.YIELD_PER)
+        )
+
+        async for dataset in stream.scalars():
+            await search_engine.delete_index(dataset)
+            await search_engine.create_index(dataset)
```

Comment:
For "soft" changes we can use the update mappings and the update settings endpoints:

See docs:
- https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html
- https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-update-settings.html

If we take care of mapping changes and we do not modify the field type, we could always update mappings without deleting the index.  See [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html#updating-field-mappings)